core/sr-primitives/src/weights.rs

```diff
@@ -0,0 +1,85 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives for transaction weighting.
+//!
+//! Each dispatch function withing `decl_module!` can now have an optional
+//! `#[weight = $x]` attribute. $x can be any object that implements the
+//! [`WeighableTransaction`] trait. By default, All transactions are annotated by
+//! `#[weight = TransactionWeight::default()]`.
+//!
+//! Note that the decl_module macro _cannot_ enforce this and will simply fail
+//! if an invalid struct is passed in.
+//!
+//! Note that [`WeighableCall`] and [`WeighableTransaction`] are more or less similar.
+//! The distinction is because one serves to pass the weight from the the
+//! dispatchable function's attribute to the call enum ([`WeighableTransaction`]) and the
+//! other to pass the final weight from call enum to the executive module
+//! ([`WeighableCall`]).
+
+/// The final type that each `#[weight = $x:expr]`'s
+/// expression must evaluate to.
+pub type Weight = u32;
+
+/// A `Call` enum that can be weighted using the custom weight attribute of the
+/// its dispatchable functions. Is implemented by default in the `decl_module!`.
+pub trait WeighableCall {
+	/// Return the weight of this call.
+	fn weight(&self, len: usize) -> Weight;
+}
+
+/// a _dispatchable_ function (anything inside `decl_module! {}`) that can be weighted.
+/// A type implementing this trait can _optionally_ be passed to the as
+/// `#[weight = X]`. Otherwise, default implementation will be used.
+pub trait WeighableTransaction {
+	/// Consume self and return the final weight of the call given the length
+	/// of the extrinsic.
+	fn calculate_weight(self, len: usize) -> Weight;
+}
+
+/// Default weight wrapper.
+/// This is tailored for the Polkadot use case. Users may replace it with anything.
+pub enum TransactionWeight {
+	/// basic weight (base, byte).
+	/// The values contained are the base weight and byte weight respectively.
+	Basic(Weight, Weight),
+	/// Maximum fee. This implies that this transaction _might_ get included but
+	/// no more transaction can be added. This can be done by setting the
+	/// implementation to _maximum block weight_.
+	Max,
+	/// Free. The transaction does not increase the total weight
+	/// (i.e. is not included in weight calculation).
+	Free,
+}
+
+impl WeighableTransaction for TransactionWeight {
+	fn calculate_weight(self, len: usize) -> Weight {
+		match self {
+			TransactionWeight::Basic(base, byte) => base + byte * len as Weight,
+			TransactionWeight::Max => 3 * 1024 * 1024,
```

Comment:
I just realised that we could calculate max given current block weight (we would have to pass it to the function along with `len`) - then you could just return whatever is left in a block.

I don't like it that much though, due to additional argument that has to be passed so if we have a better solution for longer term or we are satisfied enough with this, then let's just leave it as is.

Reply:
What you propose will actually work (haven't tried but shouldn't have a problem).

The potential long term solution is the other approach I explained [here](https://github.com/paritytech/substrate/pull/2799#issuecomment-500142494). I think in that design the weight calculation is internal to the `Module<T>` and can simply query `system`. I am still not fully convinced of how that should work (+ if it is absolutely needed) hence rather get this stable and clean first to be able to proceed to fee + tip steps and then refactor if needed (hopefully a decision I won't regret :recycle:)